### PR TITLE
Main folder: add missing API header comments #379

### PR DESCRIPTION
**Pull request by eugenepeh**
Part of #379

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@eugenepeh submitted v1 for review.

* [v1 1/1] [Main folder: add missing API header comments](https://github.com/se-edu/addressbook-level4/pull/631/commits/70e396ed61ed54ba75bc12ae7967405603f747f7)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/ddcea63541d5f73f6b3d7c6c7676d4db2bb1af9a...70e396ed61ed54ba75bc12ae7967405603f747f7))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/631/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/model/AddressBook.java [hunk 1]
```diff
@@ -60,6 +60,9 @@ public void setTags(Set<Tag> tags) {
         this.tags.setTags(tags);
     }
 
+    /**
+     * Replace existing {@code UniquePersonList} and {@code UniqueTagList} with {@code newData}'s.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/model/AddressBook.java]: This header comment is explaining the inner workings of the method :P It should simply give a high-level view of what this method is doing e.g. `Resets the existing data of this address book with {@code newData}`.

File: src/main/java/seedu/address/model/AddressBook.java [hunk 2]
```diff
@@ -139,6 +142,10 @@ private void syncMasterTagListWith(UniquePersonList persons) {
         persons.forEach(this::syncMasterTagListWith);
     }
 
+    /**
+     * Removes a person, that is equivalent to {@code key}, from {@code UniquePersonList}.
+     * @throws PersonNotFoundException if an equivalent person is not found and deleted.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/model/AddressBook.java]: Removes {@code key} from this address book. 
@throws PersonNotFoundException if {@code key} is not in this address book.

Since `UniquePersonList` is a private variable, I don't think people can see that AddressBook is implemented with a {@code UniquePersonList} anws :P

File: src/main/java/seedu/address/model/tag/UniqueTagList.java [hunk 3]
```diff
@@ -99,6 +99,9 @@ public void add(Tag toAdd) throws DuplicateTagException {
         return internalList.iterator();
     }
 
+    /**
+     * Returns the {@code internalList} as {@code UnmodifiableObservableList}.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/model/tag/UniqueTagList.java]: Same as what I mentioned above: `internalList` is a private variable :P So, do something like this instead:

`Returns the backing list as an {@code UnmodifiableObservableList}.`

File: src/main/java/seedu/address/model/tag/UniqueTagList.java [hunk 4]
```diff
@@ -112,6 +115,10 @@ public boolean equals(Object other) {
                         && this.internalList.equals(((UniqueTagList) other).internalList));
     }
 
+    /**
+     * Returns true if this {@code internalList} is equal to {@code other}'s internalList.
+     * The list order is not sensitive.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/model/tag/UniqueTagList.java]: Returns true if the elements in this ~~{@code internalList}~~ list is equal to the elements in {@code other}~~'s internalList~~.
The elements do not have to be in the same order.

File: src/main/java/seedu/address/model/util/SampleDataUtil.java [hunk 5]
```diff
@@ -14,6 +14,9 @@
 import seedu.address.model.person.exceptions.DuplicatePersonException;
 import seedu.address.model.tag.Tag;
 
+/**
+ * Contain utility methods for populating AddressBook with sample data.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java]: Contain**s** ... {@code AddressBook}

File: src/main/java/seedu/address/ui/CommandBox.java [hunk 6]
```diff
@@ -15,6 +15,9 @@
 import seedu.address.logic.commands.exceptions.CommandException;
 import seedu.address.logic.parser.exceptions.ParseException;
 
+/**
+ * The Command Box of the App.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/CommandBox.java]: I know you forked this from BrowserPanel, but we can try to be more descriptive over here :P

File: src/main/java/seedu/address/ui/CommandBox.java [hunk 7]
```diff
@@ -35,6 +38,9 @@ public CommandBox(Logic logic) {
         historySnapshot = logic.getHistorySnapshot();
     }
 
+    /**
+     * Handles the key press event, {@code keyEvent}, of the Command Box.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/CommandBox.java]: `Handles {@code keyEvent}` will do.

File: src/main/java/seedu/address/ui/CommandBox.java [hunk 8]
```diff
@@ -109,6 +118,9 @@ private void handleCommandInputChanged() {
         }
     }
 
+    /**
+     * Initializes the history snapshot.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/CommandBox.java]: I'm wondering if `Initializes {@code historySnapshot}` will do. @yamgent since this is a private method, is it ok to use `{@code}` on a private variable as well?

File: src/main/java/seedu/address/ui/HelpWindow.java [hunk 9]
```diff
@@ -39,6 +39,16 @@ public HelpWindow() {
...
+    /**
+     * Shows the help window.
+     * @throws IllegalStateException if this method is called on a thread
+     *     other than the JavaFX Application Thread.
+     * @throws IllegalStateException if this method is called during
+     *     animation or layout processing.
+     * @throws IllegalStateException if this method is called on the
+     *     primary stage.
+     * @throws IllegalStateException if {@code dialogStage} is already showing.
+     */
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/HelpWindow.java]: @damithc do we have to be so detailed in our documentation concerning unchecked exceptions?

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 10]
```diff
@@ -122,6 +122,9 @@ private void setAccelerator(MenuItem menuItem, KeyCombination keyCombination) {
         });
     }
 
+    /**
+     * Fill up all the placeholders of the Main Window.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/MainWindow.java]: Fill**s** ..... of **this window**.

File: src/main/java/seedu/address/ui/PersonCard.java [hunk 11]
```diff
@@ -8,6 +8,9 @@
 import javafx.scene.layout.Region;
 import seedu.address.model.person.ReadOnlyPerson;
 
+/**
+ * An Ui component that display information of a {@code Person}.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/PersonCard.java]: An U**I** component that display**s** the information of a {@code Person}.

File: src/main/java/seedu/address/ui/PersonListPanel.java [hunk 12]
```diff
@@ -51,6 +51,9 @@ private void setEventHandlerForSelectionChangeEvent() {
                 });
     }
 
+    /**
+     * Highlights the {@code PersonCard} at the {@code index}.
+     */
     private void scrollTo(int index) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/PersonListPanel.java]: You should mention both scrolling and selecting. :) I'm not sure whether we should use the term `highlight`.

File: src/main/java/seedu/address/ui/UiManager.java [hunk 13]
```diff
@@ -83,6 +83,9 @@ void showAlertDialogAndWait(Alert.AlertType type, String title, String headerTex
         showAlertDialogAndWait(mainWindow.getPrimaryStage(), type, title, headerText, contentText);
     }
 
+    /**
+     * Shows an alert dialog with the given parameters and wait for user response.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/UiManager.java]: Shows an alert dialog with the given parameters ~~and wait for user response~~. This method only returns after the user has closed the alert dialog.

File: src/main/java/seedu/address/ui/UiManager.java [hunk 14]
```diff
@@ -95,6 +98,9 @@ private static void showAlertDialogAndWait(Stage owner, AlertType type, String t
         alert.showAndWait();
     }
 
+    /**
+     * Shows an alert dialog with the message of {@code e} and exits the application after user confirmation.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/UiManager.java]: after ~~user confirmation~~ the user has closed the alert dialog.

File: src/main/java/seedu/address/ui/PersonListPanel.java [hunk 15]
```diff
@@ -64,6 +67,9 @@ private void handleJumpToListRequestEvent(JumpToListRequestEvent event) {
         scrollTo(event.targetIndex);
     }
 
+    /**
+     * A ListView item view that display the graphics of a {@code PersonCard}.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/PersonListPanel.java]: Custom {@code ListCell} that displays the graphics of a {@code PersonCard}.

File: src/main/java/seedu/address/MainApp.java [hunk 16]
```diff
@@ -104,6 +108,9 @@ private void initLogging(Config config) {
         LogsCenter.init(config);
     }
 
+    /**
+     * Initializes a {@code Config} with the file at {@code configFilePath}.
```
Comment (eugenepeh) [inline src/main/java/seedu/address/MainApp.java]: > Also, have to specify what happens when configFilePath is null. Maybe something like If {@code configFilePath} is null, the default file path {@code Config#DEFAULT_CONFIG_FILE} will be used.

That's what I thought initially but wouldn't that be explaining too much of the code and also the underlying implementation? There's also logging involved so I am not too sure if we should explain what if the `{@code configFilePath} is null`.

Comment (eugenepeh) [inline src/main/java/seedu/address/MainApp.java]: It's also a bit weird if the programmer be like "hmm I would like to use the DEFAULT_CONFIG_FILE" 
-> I shall use the method like this `initConfig(null);`

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/MainApp.java]: > There's also logging involved

Logging is fine because we don't mention elsewhere in the codebase in the header comments whenever we are logging. 

> explaining too much of the code

It's necessary so that developers know the behaviour of this method if they pass in a null (one of the purpose of method headers are to explain the behaviour of the method). So for e.g., `Model#updateFilteredPersonList(Predicate<ReadOnlyPerson>)` mention that a NPE will be thrown if the predicate is null.

> It's also a bit weird if the programmer be like "hmm I would like to use the DEFAULT_CONFIG_FILE" -> I shall use the method like this initConfig(null);

Haha you are perfectly right :P It should be a `initConfig()` instead actually. However, since the method is implemented this way, we have no choice but to write the appropriate header comments according to how this method is implemented currently. Unless you wanna raise PRs to fix this method ><

Comment (eugenepeh) [inline src/main/java/seedu/address/MainApp.java]: >It's necessary so that developers know the behaviour of this method if they pass in a null (one of the purpose of method headers are to explain the behaviour of the method). So for e.g., Model#updateFilteredPersonList(Predicate<ReadOnlyPerson>) mention that a NPE will be thrown if the predicate is null.

Alright, will it be better if I follow `@throw` exception style? where we put the expected outcome before the condition.

E.g.

`The default file path {@code Config#DEFAULT_CONFIG_FILE} will be used instead 
if the {@code configFilePath} is null.`

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/MainApp.java]: Yup this looks good! ^^

File: src/main/java/seedu/address/commons/core/LogsCenter.java [hunk 17]
```diff
@@ -86,6 +95,10 @@ private static void addFileHandler(Logger logger) {
         }
     }
 
+    /**
+     * Creates a new {@code fileHandler}.
+     * @throws IOException if there are problems opening the file.
```
Comment (yamgent) [inline src/main/java/seedu/address/commons/core/LogsCenter.java]: Creates a file handler for the log file.

File: src/main/java/seedu/address/ui/CommandBox.java (hunk 8 above)
Comment (yamgent) [inline src/main/java/seedu/address/ui/CommandBox.java]: I am not sure whether we are eventually converting everything to `{@link}`, but if we do, `{@link}` will not be allowing linking with variables inside the method.

Comment (eugenepeh) [inline src/main/java/seedu/address/ui/CommandBox.java]: there are some methods' header comments do that. such as `private void navigateToNextInput()`

Comment (yamgent) [inline src/main/java/seedu/address/ui/CommandBox.java]: Those have to be changed then. It is weird to reference variables inside methods, since javadocs are meant to be for the users of the API and most of the time these users won't have access to the implementation of the methods, so the variable referencing is not useful.

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/CommandBox.java]: Alright cool. That way I suppose this header comment is fine then ^^

**Comment by CanIHasReview** [comment]
# v2

@eugenepeh submitted v2 for review.

* [v2 1/1] [Main folder: add missing API header comments](https://github.com/se-edu/addressbook-level4/pull/631/commits/17e7cdb75c9bf5c87078a0a020136ec1d811bfb7)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/ddcea63541d5f73f6b3d7c6c7676d4db2bb1af9a...17e7cdb75c9bf5c87078a0a020136ec1d811bfb7)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/4ab3fdefc6aa5b13979bc9a7dab46cdbfeb496d7~1...4ab3fdefc6aa5b13979bc9a7dab46cdbfeb496d7))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/631/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/MainApp.java [hunk 18]
```diff
@@ -135,6 +145,10 @@ protected Config initConfig(String configFilePath) {
         return initializedConfig;
     }
 
+    /**
+     * Returns a {@code UserPrefs} using the file at {@code prefsFilePath} stored in {@code storage}. <br>
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/MainApp.java]: `prefsFilePath` is unknown to the reader. So something like `Returns **the** {@code UserPrefs} at {@code storage}'s user prefs file path.

File: src/main/java/seedu/address/MainApp.java [hunk 19]
```diff
@@ -135,6 +145,10 @@ protected Config initConfig(String configFilePath) {
...
+     * Returns the default {@code UserPrefs} if {@code DataConversionException} or {@code IOException} occurred.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/MainApp.java]: Same issue as above ^^ Also, perhaps we should be more specific on what's the `default {@code UserPrefs}`. Maybe something like `Returns the {@code UserPrefs} constructed by the empty {@code UserPrefs}'s constructor...` Otherwise developers will be thinking "What's the default {@code UserPrefs}?"

File: src/main/java/seedu/address/commons/core/LogsCenter.java [hunk 20]
```diff
@@ -61,20 +61,31 @@ public static Logger getLogger(String name) {
...
         if (consoleHandler == null) {
             consoleHandler = createConsoleHandler();
         }
         logger.addHandler(consoleHandler);
     }
 
+    /**
+     * Remove all the handlers from {@code logger}.
+     */
     private static void removeHandlers(Logger logger) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/commons/core/LogsCenter.java]: If you update this method to use lambda: `Arrays.stream(handlers).forEach(logger::removeHandler);`

We can remove the header comments :P. That needs to be in another PR though.

File: src/main/java/seedu/address/model/util/SampleDataUtil.java [hunk 21]
```diff
@@ -14,6 +14,9 @@
 import seedu.address.model.person.exceptions.DuplicatePersonException;
 import seedu.address.model.tag.Tag;
 
+/**
+ * Contains utility methods for populating {@code AddressBook} with sample data.
+ */
 public class SampleDataUtil {
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java]: just curious, public static methods don't require header comments?

File: src/main/java/seedu/address/ui/UiManager.java [hunk 22]
```diff
@@ -83,6 +83,10 @@ void showAlertDialogAndWait(Alert.AlertType type, String title, String headerTex
         showAlertDialogAndWait(mainWindow.getPrimaryStage(), type, title, headerText, contentText);
     }
 
+    /**
+     * Shows an alert dialog with the given parameters.
+     * This method only returns after the user has closed the alert dialog.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/UiManager.java]: I think the header comments can be more specific. Maybe something like: `Shows an alert dialog on {@code owner} with the given parameters.`

File: src/main/java/seedu/address/ui/UiManager.java [hunk 23]
```diff
@@ -95,6 +99,10 @@ private static void showAlertDialogAndWait(Stage owner, AlertType type, String t
         alert.showAndWait();
     }
 
+    /**
+     * Shows an alert dialog with the message of {@code e} and exits the application after
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/UiManager.java]: Can be more specific too. `Shows an **error** alert dialog that displays {@code title} and the error message of {@code e}, and...`

File: src/main/java/seedu/address/model/util/SampleDataUtil.java (hunk 21 above)
Comment (eugenepeh) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java]: its not that public static don't require, its the getters and setters :P

File: src/main/java/seedu/address/commons/core/LogsCenter.java (hunk 20 above)
Comment (eugenepeh) [inline src/main/java/seedu/address/commons/core/LogsCenter.java]: True but if we do that, we need to fix all the others (if there's any) as well :P
Have any of the addressbook's code started using lambda yet?

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/commons/core/LogsCenter.java]: > True but if we do that, we need to fix all the others (if there's any) as well :P

I would imagine this to be a really boring PR ><

> Have any of the addressbook's code started using lambda yet?

I converted a lot of the code to lambda expressions actually :P

File: src/main/java/seedu/address/model/util/SampleDataUtil.java (hunk 21 above)
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java]: Ah I see. Some public getter methods are complex though, for e.g. those in `SampleDataUtil` class, and perhaps having a header comment that would be helpful. Do you think we should raise an issue to improve the Checkstyle, since I think you're done with your 2103R right :P

Comment (eugenepeh) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java]: I believe, it is still possible to do something like that by adding another method rule that does not ignore getters and setters and set a different length limit e.g. 5. But it's a very gray area where the programmer may deems that there is no more useful information which he can add to the header comment. So in the end checkstyle only can assist but not foolproof. :P

>I think you're done with your 2103R right i.e. this is your last PR or smth like that :P

Pretty much :P but i will still continue to contribute from time to time

File: src/main/java/seedu/address/commons/core/LogsCenter.java (hunk 20 above)
Comment (eugenepeh) [inline src/main/java/seedu/address/commons/core/LogsCenter.java]: Opened an issue for this #634 but let's get this PR approved first before I work on that one :P

File: src/main/java/seedu/address/MainApp.java (hunk 19 above)
Comment (eugenepeh) [inline src/main/java/seedu/address/MainApp.java]: Hmm empty constructor sounds weird though because it does comes with the default `guiSettings`, `addressBookFilePath` and `addressBookName`. 
How about `Returns a new {@code UserPrefs} with default configuration`?

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/MainApp.java]: Yup sounds good! :+1:

**Comment by Zhiyuan-Amos** [comment]
Just curious @eugenepeh, why didn't you merge this PR with #379? You can have the first commit whereby you update all the header comments (i.e. this PR) and in the next commit you can update the checkstyle? That way you can be sure that you didn't miss anything out isn't it :P

**Comment by eugenepeh** [comment]
@Zhiyuan-Amos it was too big initially and along with all ongoing PRs, most likely it will end up with a lot of collisions :P. Other contributors suggested that I should break it into smaller PRs, so it will be easier on the reviewers.

**Comment by Zhiyuan-Amos** [comment]
@eugenepeh ah I see. Since we won't be merging any PRs soon, feel free to combine this PR with #379 if it makes your work easier ^^

**Comment by eugenepeh** [comment]
continue in #379
